packages/labs/src/frameworks/react/create-component.ts

```diff
@@ -0,0 +1,185 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as ReactModule from 'react';
+
+const reservedReactProperties = new Set([
+  'children',
+  'localName',
+  'ref',
+  // TODO(sorvell): why are the properties below included?
+  'style',
+  'className',
+]);
+
+const setProperty = <E extends Element, T>(
+  node: E,
+  name: string,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  value: any,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  old: any,
+  events?: T
+) => {
+  // Dirty check and prevent setting reserved properties.
+  if (Object.is(value, old) || reservedReactProperties.has(name as string)) {
+    return;
+  }
+  // For events, use an explicit list.
+  const event = (events?.[
+    name as keyof T
+  ] as unknown) as keyof HTMLElementEventMap;
+  if (event !== undefined) {
+    if (old !== undefined) {
+      node.removeEventListener(event, old);
+    }
+    node.addEventListener(event, value);
+  } else {
+    // For properties, use an `in` check
+    if (name in node) {
+      node[name as keyof E] = value;
+      // And for everything else, set the attribute.
+    } else if (value == null) {
+      node.removeAttribute(name);
+    } else {
+      node.setAttribute(name, value);
+    }
+  }
+};
+
+type ForwardedRef = {
+  __forwardedRef?: React.Ref<unknown>;
+};
+
+type Events<S> = {
+  [P in keyof S]?: (e: Event) => unknown;
+};
+
+const componentMap: Map<
+  string,
+  React.ForwardRefExoticComponent<
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    React.PropsWithChildren<React.PropsWithRef<any>>
+  >
+> = new Map();
+
+/**
+ *  Creates a React component from a CustomElement.
+ */
+// TODO(sorvell): Was unable to remove type E here.
+export const createComponent = <T extends HTMLElement, E>(
+  React: typeof ReactModule,
+  tagName: string,
+  events?: E
```

Comment:
Events is very loosely typed this way - effectively `unknown`, so the generic doesn't buy you much. I'd type as at least `object` or `extends object`, or better an indexed type with a string value.

Reply:
Is there a way of typing `E` as requiring its members to be of type `string` without using an `index` signature? The idea is that we only want the fields passed via `E` to be acceptable `as props`.

Reply:
I'm not sure I follow. Prop keys can be strings, so that's the right type, no?